receiver.go

```diff
@@ -0,0 +1,250 @@
+		rmCallCh:  rmCallCh,
+		stopCh:    make(chan struct{}),
+		closedCh:  make(chan struct{}),
+
+		taskBeginCh:  make(chan *task),
+		taskCancelCh: make(chan int),
+		taskEndCh:    make(chan int),
+
+		log:           l,
+		wrapErrorFunc: wef,
+	}
+	r.messageHandlers = map[MethodType]messageHandler{
+		MethodNotify:   {dispatchFunc: r.receiveNotify, messageLength: 3},
+		MethodCall:     {dispatchFunc: r.receiveCall, messageLength: 4},
+		MethodResponse: {dispatchFunc: r.receiveResponse, messageLength: 4},
+		MethodCancel:   {dispatchFunc: r.receiveCancel, messageLength: 3},
+	}
+	go r.taskLoop()
+	return r
+}
+
+func (r *receiveHandler) taskLoop() {
+	tasks := make(map[int]context.CancelFunc)
+	for {
+		select {
+		case <-r.stopCh:
+			close(r.closedCh)
+			return
+		case t := <-r.taskBeginCh:
+			tasks[t.seqid] = t.cancelFunc
+		case seqid := <-r.taskCancelCh:
+			if cancelFunc, ok := tasks[seqid]; ok {
+				cancelFunc()
+			}
+			delete(tasks, seqid)
+		case seqid := <-r.taskEndCh:
+			delete(tasks, seqid)
+		}
+	}
+}
+
+func (r *receiveHandler) findServeHandler(n string) (*ServeHandlerDescription, WrapErrorFunc, error) {
+	p, m := SplitMethodName(n)
+	prot, found := r.protocols[p]
+	if !found {
+		return nil, r.wrapErrorFunc, ProtocolNotFoundError{p}
+	}
+	srv, found := prot.Methods[m]
+	if !found {
+		return nil, r.wrapErrorFunc, MethodNotFoundError{p, m}
+	}
+	return &srv, prot.WrapError, nil
+}
+
+func (d *receiveHandler) Receive(length int) error {
+	var requestType MethodType
+	if err := d.reader.Decode(&requestType); err != nil {
+		return err
+	}
+	handler, ok := d.messageHandlers[requestType]
+	if !ok {
+		return NewDispatcherError("invalid message type")
+	}
+	if length != handler.messageLength {
+		return NewDispatcherError("wrong number of fields for message (got n=%d, expected n=%d)", length, handler.messageLength)
+
+	}
+	return handler.dispatchFunc()
+}
+
+func (r *receiveHandler) receiveNotify() (err error) {
+	req := newRequest(MethodNotify)
+	return r.handleReceiveDispatch(req)
+}
+
+func (r *receiveHandler) receiveCall() error {
+	req := newRequest(MethodCall)
+	return r.handleReceiveDispatch(req)
+}
+
+func (r *receiveHandler) receiveCancel() (err error) {
+	req := newRequest(MethodCancel)
+	if err := decodeIntoRequest(r.reader, req); err != nil {
+		return err
+	}
+	req.LogInvocation(r.log, nil, nil)
+	r.taskCancelCh <- req.Message().seqno
+	return nil
+}
+
+func (r *receiveHandler) handleReceiveDispatch(req request) error {
+	if err := decodeIntoRequest(r.reader, req); err != nil {
+		return err
+	}
+
+	m := req.Message()
+	serveHandler, wrapErrorFunc, se := r.findServeHandler(m.method)
+	if se != nil {
+		m.err = wrapError(wrapErrorFunc, se)
+		if err := decodeToNull(r.reader, m); err != nil {
+			return err
+		}
+		req.LogInvocation(r.log, se, nil)
+		return req.Reply(r.writer, r.log)
+	}
+	cancelFunc := req.Serve(r.reader, r.writer, serveHandler, wrapErrorFunc, r.log)
+	if cancelFunc != nil {
+
```

Comment:
OK I refactored it as part of my test update. I suggest we merge this in because the above code doesn't quite yet get exercised by this change. The next PR makes cancel synchronous The cancel itself is async but the canceled call doesn't return until the cancellation is dispatched and handled. This means that you can control what is returned upon a cancellation.